### ISSUE TITLE
`forever: true` causes memory to leak like the deepwater horizon

### ISSUE THREAD

**Issue by devTristan** [request/request#1637]
```javascript
var request = require('request');

var concurrency = 128;
var inProgress = 0;
var startedCount = 0;
var finishedCount = 0;

function fillQueue() {
  while (inProgress < concurrency) {
    makeRequest();
  }
}

function makeRequest() {
  inProgress++;
  startedCount++;

  request.post({
    uri: 'http://localhost/',
    forever: true
  }, function(){
    inProgress--;
    finishedCount++;
    fillQueue();
  });
}

fillQueue();

setInterval(function(){
  console.log(inProgress+' requests in progress, '+startedCount+' started, '+finishedCount+' finished');
}, 1000);
```

When executed, this script will leak memory until you run out of it. This effect has been observed on ubuntu 14.04 using node 0.11.16 and 0.12.4. **There is no memory leak in 0.10.25**.

Don't run this script if you have a server running locally on port 80, this might DOS it.

Commenting out `forever: true` will cause this script to run as normal, but without the memory leak, and presumably without keepalive support.

**Comment by seamusabshere** [comment]
@devTristan do you know what is going on?

**Comment by devTristan** [comment]
My first guess would be something in [here](https://github.com/request/forever-agent/blob/master/index.js) preventing request data from being garbage-collected. Very possible that I'm wrong, though.

**Comment by devTristan** [comment]
@simov gentle ping

**Comment by mikeal** [comment]
0.12+ makes this feature totally unnecessary. We should check if the node version is greater than 0.10.x and print a warning when using this feature and silently turn it off.

**Comment by simov** [comment]
:+1: I was just trying to figure out what's the difference between forever-agent and the current Agent implementation in core.

**Comment by simov** [comment]
@devTristan, theoretically this should give you the same results, without the memory leak. Let me know what do you think.

```js
var request = require('request');
var http = require('http');

var concurrency = 128;
var inProgress = 0;
var startedCount = 0;
var finishedCount = 0;

function fillQueue() {
  while (inProgress < concurrency) {
    makeRequest();
  }
}

function makeRequest() {
  inProgress++;
  startedCount++;

  request.post({
    uri: 'http://localhost/',
    agentOptions: {keepAlive:true, maxSockets:1},
    agentClass: http.Agent
    // forever: true
  }, function(){
    inProgress--;
    finishedCount++;
    fillQueue();
  });
}

fillQueue();

setInterval(function(){
  console.log(inProgress+' requests in progress, '+startedCount+' started, '+finishedCount+' finished');
}, 1000);
```

**Comment by devTristan** [comment]
It does indeed. Same results, no memory leak.

**Comment by simov** [comment]
Ok, I should probably play around with these options and figure out some fix, as mikeal pointed out. Our docs should be improved too.

**Comment by simov** [comment]
btw `{keepAlive:true, maxSockets:1}` closes the socket in `node 0.12.6` even if you specify big enough `keepAliveMsecs` timeout, works in `iojs` though :astonished:
